R/staticServer.R

```diff
@@ -0,0 +1,80 @@
+#' Serve a directory
+#'
+#' Run a server to host a static directory.
+#'
+#' @examplesIf interactive()
+#' website_dir <- system.file("example-static-site", package = "httpuv")
+#' runStaticServer(dir = website_dir)
+#'
+#' @param dir The directory to serve. Defaults to the current working directory.
+#' @inheritParams startServer
+#' @param background Whether to run the server in the background. By default,
+#'   the server runs in the foreground and blocks the R console. You can stop
+#'   the server by interrupting it with `Ctrl + C`.
+#'
+#'   When `background = TRUE`, the server will run in the background and will
+#'   process requests when the R console is idle. To stop a background server,
+#'   call [stopAllServers()] or call [stopServer()] on the server object
+#'   returned (invisibly) by this function.
+#'
+#' @seealso [runServer()] provides a similar interface for running a dynamic
+#'   app server. Both `runStaticServer()` and [runServer()] are built on top of
+#'   [startServer()], [service()] and [stopServer()]. Learn more about httpuv
+#'   servers in [startServer()].
+#'
+#' @export
+runStaticServer <- function(
+  dir = getwd(),
+  host = "127.0.0.1",
+  port = 7446,
+  background = FALSE,
+  browse = interactive()
+) {
+  stopifnot(
+    "`port` must be an integer" = is.numeric(port),
+    "`port` must be an integer" = port == as.integer(port),
+    "`port` must be a single integer" = length(port) == 1
+  )
+
+  browse <- isTRUE(browse)
+  port <- requested_or_random_port(port, host)
+  root <- staticPath(dir)
+
+  server <- startServer(
+    app = list(staticPaths = list("/" = root)),
+    host = host,
+    port = port
+  )
+
+  message("Serving: '", dir, "'")
+  message("View at: http://", host, ":", port, sep = "")
+
+  if (browse) {
+    utils::browseURL(paste0("http://", host, ":", port))
+  }
+
+  if (background) return(invisible(server))
+
+  on.exit(stopServer(server))
+  message("Press Esc or Ctrl + C to stop the server")
+  service(0)
+}
+
+requested_or_random_port <- function(port, host = "127.0.0.1") {
+  tryCatch(
+    # Is the requested port open?
+    randomPort(port, port, host = host),
+    error = function(err) {
+      msg <- conditionMessage(err)
+      if (!grepl("Cannot find an available port", msg, fixed = TRUE)) {
+        stop(msg)
```

Comment:
If the intention is to re-raise the error, then how about:
```suggestion
        stop(err)
```

Reply:
I refactored the internals of `randomPort()` to throw an error with a custom class (f5d699760ff01de7fde19f91a8e82fd353889a66) and then pulled out the try and throw logic into smaller helper functions in a637e085c88a1a95ce89386b3a33412c3e24e9b9.

Breaking apart the logic let me be more prescriptive about the default port handling and I ultimately didn't need the custom error class in this PR. But I've run into situations where a classed error would be helpful and think it's worth keeping